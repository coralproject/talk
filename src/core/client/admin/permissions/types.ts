import { GQLUSER_ROLE_RL } from "coral-framework/schema";

export type PermissionMap<AbilityType extends string, ContextType> = {
  [abilityType in AbilityType]: {
    [role in GQLUSER_ROLE_RL]?: (context?: ContextType) => boolean;
  };
};
