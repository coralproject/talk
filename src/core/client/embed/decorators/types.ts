import pym from "pym.js";

export type CleanupCallback = () => void;
export type Decorator = (pym: pym.Parent) => CleanupCallback | void;
