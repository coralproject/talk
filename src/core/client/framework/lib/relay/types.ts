import { _RefType } from "react-relay";

export type FragmentKeys<T> = {
  [P in keyof T]: T[P] extends _RefType<any> | null ? P : never
}[keyof T];

export type FragmentKeysNoLocal<T> = Exclude<FragmentKeys<T>, "local">;
