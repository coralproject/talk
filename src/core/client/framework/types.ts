// TODO: (@cvle) Extract useful common types into its own package.
export { Diff, Omit, Overwrite, PropTypesOf } from "talk-ui/types";
