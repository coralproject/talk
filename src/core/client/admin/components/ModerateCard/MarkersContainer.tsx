import { Localized } from "@fluent/react/compat";
import React, { useMemo } from "react";
import { graphql } from "react-relay";

import { withFragmentContainer } from "coral-framework/lib/relay";
import { Marker, MarkerCount } from "coral-ui/components/v2";

import { MarkersContainer_comment } from "coral-admin/__generated__/MarkersContainer_comment.graphql";
import { MarkersContainer_settings } from "coral-admin/__generated__/MarkersContainer_settings.graphql";

import Markers from "./Markers";
import ModerateCardDetailsContainer from "./ModerateCardDetailsContainer";

interface MarkersContainerProps {
  comment: MarkersContainer_comment;
  onUsernameClick: (id?: string) => void;
  settings: MarkersContainer_settings;
}

let keyCounter = 0;
const markers: Array<(
  c: MarkersContainer_comment
) => React.ReactElement<any> | null> = [
  (c) =>
    (c.status === "PREMOD" && (
      <Localized id="moderate-marker-preMod" key={keyCounter++}>
        <Marker color="pending">Pre-Mod</Marker>
      </Localized>
    )) ||
    null,
  (c) =>
    (c.revision &&
      c.revision.actionCounts.flag.reasons.COMMENT_DETECTED_LINKS && (
        <Localized id="moderate-marker-link" key={keyCounter++}>
          <Marker color="pending">Link</Marker>
        </Localized>
      )) ||
    null,
  (c) =>
    (c.revision &&
      c.revision.actionCounts.flag.reasons.COMMENT_DETECTED_BANNED_WORD &&
      c.revision.metadata?.wordList?.timedOut && (
        <Localized id="moderate-marker-possibleBannedWord" key={keyCounter++}>
          <Marker color="reported">Possible Banned Word</Marker>
        </Localized>
      )) ||
    null,
  (c) =>
    (c.revision &&
      c.revision.actionCounts.flag.reasons.COMMENT_DETECTED_BANNED_WORD &&
      !c.revision.metadata?.wordList?.timedOut && (
        <Localized id="moderate-marker-bannedWord" key={keyCounter++}>
          <Marker color="reported">Banned Word</Marker>
        </Localized>
      )) ||
    null,
  (c) =>
    (c.revision &&
      c.revision.actionCounts.flag.reasons.COMMENT_DETECTED_SUSPECT_WORD &&
      c.revision.metadata?.wordList?.timedOut && (
        <Localized id="moderate-marker-possibleSuspectWord" key={keyCounter++}>
          <Marker color="reported">Possible Suspect Word</Marker>
        </Localized>
      )) ||
    null,
  (c) =>
    (c.revision &&
      c.revision.actionCounts.flag.reasons.COMMENT_DETECTED_SUSPECT_WORD &&
      !c.revision.metadata?.wordList?.timedOut && (
        <Localized id="moderate-marker-suspectWord" key={keyCounter++}>
          <Marker color="reported">Suspect Word</Marker>
        </Localized>
      )) ||
    null,
  (c) =>
    (c.revision &&
      c.revision.actionCounts.flag.reasons.COMMENT_DETECTED_SPAM && (
        <Localized id="moderate-marker-spamDetected" key={keyCounter++}>
          <Marker color="reported">Spam Detected</Marker>
        </Localized>
      )) ||
    null,
  (c) =>
    (c.revision &&
      c.revision.actionCounts.flag.reasons.COMMENT_DETECTED_TOXIC && (
        <Localized id="moderate-marker-toxic" key={keyCounter++}>
          <Marker color="reported">Toxic</Marker>
        </Localized>
      )) ||
    null,
  (c) =>
    (c.revision &&
      c.revision.actionCounts.flag.reasons.COMMENT_DETECTED_REPEAT_POST && (
        <Localized id="moderate-marker-repeatPost" key={keyCounter++}>
          <Marker color="reported">Repeat comment</Marker>
        </Localized>
      )) ||
    null,
  (c) =>
    (c.revision &&
      c.revision.actionCounts.flag.reasons.COMMENT_DETECTED_RECENT_HISTORY && (
        <Localized id="moderate-marker-recentHistory" key={keyCounter++}>
          <Marker color="reported">Recent History</Marker>
        </Localized>
      )) ||
    null,
  (c) =>
    (c.revision &&
      c.revision.actionCounts.flag.reasons.COMMENT_REPORTED_OFFENSIVE && (
        <Marker key={keyCounter++} color="reported">
          <Localized id="moderate-marker-offensive">
            <span>Offensive</span>
          </Localized>{" "}
          <MarkerCount>
            {c.revision.actionCounts.flag.reasons.COMMENT_REPORTED_OFFENSIVE}
          </MarkerCount>
        </Marker>
      )) ||
    null,
  (c) =>
    (c.revision &&
      c.revision.actionCounts.flag.reasons.COMMENT_REPORTED_ABUSIVE && (
        <Marker key={keyCounter++} color="reported">
          <Localized id="moderate-marker-abusive">
            <span>Abusive</span>
          </Localized>{" "}
          <MarkerCount>
            {c.revision.actionCounts.flag.reasons.COMMENT_REPORTED_ABUSIVE}
          </MarkerCount>
        </Marker>
      )) ||
    null,
  (c) =>
    (c.revision &&
      c.revision.actionCounts.flag.reasons.COMMENT_REPORTED_SPAM && (
        <Marker key={keyCounter++} color="reported">
          <Localized id="moderate-marker-spam">
            <span>Spam</span>
          </Localized>{" "}
          <MarkerCount>
            {c.revision.actionCounts.flag.reasons.COMMENT_REPORTED_SPAM}
          </MarkerCount>
        </Marker>
      )) ||
    null,
  (c) =>
    (c.revision &&
      c.revision.actionCounts.flag.reasons.COMMENT_REPORTED_OTHER && (
        <Marker key={keyCounter++} color="reported">
          <Localized id="moderate-marker-other">
            <span>Other</span>
          </Localized>{" "}
          <MarkerCount>
            {c.revision.actionCounts.flag.reasons.COMMENT_REPORTED_OTHER}
          </MarkerCount>
        </Marker>
      )) ||
    null,
  (c) =>
    (c.revision &&
      c.revision.actionCounts.flag.reasons.COMMENT_DETECTED_NEW_COMMENTER && (
        <Localized id="moderate-marker-newCommenter" key={keyCounter++}>
          <Marker color="reported">New commenter</Marker>
        </Localized>
      )) ||
    null,
];

export const MarkersContainer: React.FunctionComponent<MarkersContainerProps> = (
  props
) => {
  const elements = useMemo(
    () => markers.map((cb) => cb(props.comment)).filter((m) => m),
    [props.comment]
  );

  return (
    <Markers
      details={
        <ModerateCardDetailsContainer
          onUsernameClick={props.onUsernameClick}
          comment={props.comment}
          settings={props.settings}
        />
      }
    >
      {elements}
    </Markers>
  );
};

const enhanced = withFragmentContainer<MarkersContainerProps>({
  comment: graphql`
    fragment MarkersContainer_comment on Comment {
      ...ModerateCardDetailsContainer_comment
      status
      revision {
        actionCounts {
          flag {
            reasons {
              COMMENT_REPORTED_SPAM
              COMMENT_REPORTED_OTHER
              COMMENT_REPORTED_OFFENSIVE
              COMMENT_REPORTED_ABUSIVE
              COMMENT_DETECTED_TOXIC
              COMMENT_DETECTED_SUSPECT_WORD
              COMMENT_DETECTED_SPAM
              COMMENT_DETECTED_REPEAT_POST
              COMMENT_DETECTED_RECENT_HISTORY
              COMMENT_DETECTED_NEW_COMMENTER
              COMMENT_DETECTED_LINKS
              COMMENT_DETECTED_BANNED_WORD
            }
          }
        }
        metadata {
          wordList {
            timedOut
          }
        }
      }
    }
  `,
  settings: graphql`
    fragment MarkersContainer_settings on Settings {
      ...ModerateCardDetailsContainer_settings
    }
  `,
})(MarkersContainer);

export default enhanced;
