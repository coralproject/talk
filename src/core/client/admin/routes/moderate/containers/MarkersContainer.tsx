import React from "react";
import { graphql } from "react-relay";

import { MarkersContainer_comment as CommentData } from "talk-admin/__generated__/MarkersContainer_comment.graphql";
import { withFragmentContainer } from "talk-framework/lib/relay";
import { Marker, MarkerCount } from "talk-ui/components";

interface MarkersContainerProps {
  comment: CommentData;
}

let keyCounter = 0;
const markers: Array<(c: CommentData) => React.ReactElement<any> | null> = [
  c =>
    (c.status === "PREMOD" && (
      <Marker key={keyCounter++} color="primary">
        Pre-Mod
      </Marker>
    )) ||
    null,
  c =>
    (c.actionCounts.flag.reasons.COMMENT_DETECTED_LINKS && (
      <Marker key={keyCounter++} color="primary">
        Link
      </Marker>
    )) ||
    null,
  c =>
    (c.actionCounts.flag.reasons.COMMENT_DETECTED_BANNED_WORD && (
      <Marker key={keyCounter++} color="error">
        Banned Word
      </Marker>
    )) ||
    null,
  c =>
    (c.actionCounts.flag.reasons.COMMENT_DETECTED_SUSPECT_WORD && (
      <Marker key={keyCounter++} color="error" variant="filled">
        Suspect Word
      </Marker>
    )) ||
    null,
  c =>
    (c.actionCounts.flag.reasons.COMMENT_DETECTED_SPAM && (
      <Marker key={keyCounter++} color="error">
        Spam
      </Marker>
    )) ||
    null,
  c =>
    (c.actionCounts.flag.reasons.COMMENT_DETECTED_TOXIC && (
      <Marker key={keyCounter++} color="error">
        Toxic
      </Marker>
    )) ||
    null,
  c =>
    (c.actionCounts.flag.reasons.COMMENT_DETECTED_TRUST && (
      <Marker key={keyCounter++} color="error">
        Karma
      </Marker>
    )) ||
    null,
  c =>
    (c.actionCounts.flag.reasons.COMMENT_DETECTED_BODY_COUNT && (
      <Marker key={keyCounter++} color="error">
        Body Count
      </Marker>
    )) ||
    null,
  c =>
    (c.actionCounts.flag.reasons.COMMENT_REPORTED_OFFENSIVE && (
      <Marker key={keyCounter++} color="error">
        Offensive{" "}
        <MarkerCount>
          {c.actionCounts.flag.reasons.COMMENT_REPORTED_OFFENSIVE}
        </MarkerCount>
      </Marker>
    )) ||
    null,
  c =>
    (c.actionCounts.flag.reasons.COMMENT_REPORTED_SPAM && (
      <Marker key={keyCounter++} color="error">
        Spam{" "}
        <MarkerCount>
          {c.actionCounts.flag.reasons.COMMENT_REPORTED_SPAM}
        </MarkerCount>
      </Marker>
    )) ||
    null,
];

class MarkersContainer extends React.Component<MarkersContainerProps> {
  public render() {
    return markers.map(cb => cb(this.props.comment));
  }
}

const enhanced = withFragmentContainer<MarkersContainerProps>({
  comment: graphql`
    fragment MarkersContainer_comment on Comment {
      status
      actionCounts {
        flag {
          reasons {
            COMMENT_DETECTED_TOXIC
            COMMENT_DETECTED_SPAM
            COMMENT_DETECTED_BODY_COUNT
            COMMENT_DETECTED_TRUST
            COMMENT_DETECTED_LINKS
            COMMENT_DETECTED_BANNED_WORD
            COMMENT_DETECTED_SUSPECT_WORD
            COMMENT_REPORTED_OFFENSIVE
            COMMENT_REPORTED_SPAM
          }
        }
      }
    }
  `,
})(MarkersContainer);

export default enhanced;
