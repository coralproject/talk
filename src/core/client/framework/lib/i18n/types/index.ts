export { default as FluentShortNumber } from "./FluentShortNumber";
