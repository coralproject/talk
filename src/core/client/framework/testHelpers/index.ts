export {
  default as createRelayEnvironment,
  CreateRelayEnvironmentParams,
} from "./createRelayEnvironment";
export { default as createFluentBundle } from "./createFluentBundle";
export { default as createSinonStub } from "./createSinonStub";
export {
  default as removeFragmentRefs,
  NoFragmentRefs,
} from "./removeFragmentRefs";
export { default as createUUIDGenerator } from "./createUUIDGenerator";
export * from "./denormalize";
export { default as replaceHistoryLocation } from "./replaceHistoryLocation";
