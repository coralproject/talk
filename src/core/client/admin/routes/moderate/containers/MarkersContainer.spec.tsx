import React from "react";
import { createRenderer } from "react-test-renderer/shallow";

import { removeFragmentRefs } from "talk-framework/testHelpers";
import { PropTypesOf } from "talk-framework/types";

import { MarkersContainer } from "./MarkersContainer";

const MarkersContainerN = removeFragmentRefs(MarkersContainer);

it("renders all markers", () => {
  const props: PropTypesOf<typeof MarkersContainerN> = {
    comment: {
      status: "PREMOD",
      actionCounts: {
        flag: {
          reasons: {
            COMMENT_DETECTED_TOXIC: 1,
            COMMENT_DETECTED_SPAM: 1,
            COMMENT_DETECTED_TRUST: 1,
            COMMENT_DETECTED_LINKS: 1,
            COMMENT_DETECTED_BANNED_WORD: 1,
            COMMENT_DETECTED_SUSPECT_WORD: 1,
            COMMENT_REPORTED_OFFENSIVE: 2,
            COMMENT_REPORTED_SPAM: 3,
          },
        },
      },
    },
  };
  const renderer = createRenderer();
  renderer.render(<MarkersContainerN {...props} />);
  expect(renderer.getRenderOutput()).toMatchSnapshot();
});

it("renders some markers", () => {
  const props: PropTypesOf<typeof MarkersContainerN> = {
    comment: {
      status: "PREMOD",
      actionCounts: {
        flag: {
          reasons: {
            COMMENT_DETECTED_TOXIC: 1,
            COMMENT_DETECTED_SPAM: 0,
            COMMENT_DETECTED_TRUST: 1,
            COMMENT_DETECTED_LINKS: 0,
            COMMENT_DETECTED_BANNED_WORD: 1,
            COMMENT_DETECTED_SUSPECT_WORD: 0,
            COMMENT_REPORTED_OFFENSIVE: 2,
            COMMENT_REPORTED_SPAM: 0,
          },
        },
      },
    },
  };
  const renderer = createRenderer();
  renderer.render(<MarkersContainerN {...props} />);
  expect(renderer.getRenderOutput()).toMatchSnapshot();
});
