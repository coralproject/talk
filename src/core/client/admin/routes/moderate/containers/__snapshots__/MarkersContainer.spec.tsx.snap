// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders all markers 1`] = `
Array [
  <Localized
    id="moderate-marker-preMod"
  >
    <withPropsOnChange(Marker)
      color="primary"
    >
      Pre-Mod
    </withPropsOnChange(Marker)>
  </Localized>,
  <Localized
    id="moderate-marker-link"
  >
    <withPropsOnChange(Marker)
      color="primary"
    >
      Link
    </withPropsOnChange(Marker)>
  </Localized>,
  <Localized
    id="moderate-marker-bannedWord"
  >
    <withPropsOnChange(Marker)
      color="error"
    >
      Banned Word
    </withPropsOnChange(Marker)>
  </Localized>,
  <Localized
    id="moderate-marker-suspectWord"
  >
    <withPropsOnChange(Marker)
      color="error"
      variant="filled"
    >
      Suspect Word
    </withPropsOnChange(Marker)>
  </Localized>,
  <Localized
    id="moderate-marker-spam"
  >
    <withPropsOnChange(Marker)
      color="error"
    >
      Spam
    </withPropsOnChange(Marker)>
  </Localized>,
  <Localized
    id="moderate-marker-toxic"
  >
    <withPropsOnChange(Marker)
      color="error"
    >
      Toxic
    </withPropsOnChange(Marker)>
  </Localized>,
  <Localized
    id="moderate-marker-karma"
  >
    <withPropsOnChange(Marker)
      color="error"
    >
      Karma
    </withPropsOnChange(Marker)>
  </Localized>,
  <withPropsOnChange(Marker)
    color="error"
  >
    <Localized
      id="moderate-marker-offensive"
    >
      <span>
        Offensive
      </span>
    </Localized>
     
    <withPropsOnChange(Count)>
      2
    </withPropsOnChange(Count)>
  </withPropsOnChange(Marker)>,
  <withPropsOnChange(Marker)
    color="error"
  >
    <Localized
      id="moderate-marker-spam"
    >
      <span>
        Spam
      </span>
    </Localized>
     
    <withPropsOnChange(Count)>
      3
    </withPropsOnChange(Count)>
  </withPropsOnChange(Marker)>,
]
`;

exports[`renders some markers 1`] = `
Array [
  <Localized
    id="moderate-marker-preMod"
  >
    <withPropsOnChange(Marker)
      color="primary"
    >
      Pre-Mod
    </withPropsOnChange(Marker)>
  </Localized>,
  null,
  <Localized
    id="moderate-marker-bannedWord"
  >
    <withPropsOnChange(Marker)
      color="error"
    >
      Banned Word
    </withPropsOnChange(Marker)>
  </Localized>,
  null,
  null,
  <Localized
    id="moderate-marker-toxic"
  >
    <withPropsOnChange(Marker)
      color="error"
    >
      Toxic
    </withPropsOnChange(Marker)>
  </Localized>,
  <Localized
    id="moderate-marker-karma"
  >
    <withPropsOnChange(Marker)
      color="error"
    >
      Karma
    </withPropsOnChange(Marker)>
  </Localized>,
  <withPropsOnChange(Marker)
    color="error"
  >
    <Localized
      id="moderate-marker-offensive"
    >
      <span>
        Offensive
      </span>
    </Localized>
     
    <withPropsOnChange(Count)>
      2
    </withPropsOnChange(Count)>
  </withPropsOnChange(Marker)>,
  null,
]
`;
